invenio_curations/services/service.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (C) 2024 CERN.
+#
+# Invenio-Curations is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+"""Curation service."""
+from flask import current_app
+from invenio_accounts.models import Role
+from invenio_i18n import gettext as _
+from invenio_records_resources.services.uow import unit_of_work
+from invenio_requests.proxies import current_request_type_registry
+from invenio_requests.services.results import ResolverRegistry
+from invenio_search.engine import dsl
+
+from ..requests import CurationRequest
+from .errors import OpenRecordCurationRequestAlreadyExists
+
+
+class CurationRequestService:
+    """Service for User Moderation requests."""
+
+    def __init__(self, requests_service, **kwargs):
+        """Service initialisation as a sub-service of requests."""
+        self.requests_service = requests_service
+
+    def role(self):
+        """Creates a RDMCuration request and submits it."""
+        role_name = current_app.config.get("CURATIONS_MODERATION_ROLE")
+        return Role.query.filter(Role.name == role_name).one_or_none()
+
+    @property
+    def request_type_cls(self):
+        """User moderation request type."""
+        return current_request_type_registry.lookup(CurationRequest.type_id)
+
+    def get_review(self, identity, topic, **kwargs):
+        topic_reference = ResolverRegistry.reference_entity(topic)
+        results = self.requests_service.search(
+            identity,
+            extra_filter=dsl.query.Bool(
+                "must",
+                must=[
+                    dsl.Q("term", **{"type": self.request_type_cls.type_id}),
+                    # TODO: make type independent search
+                    dsl.Q("term", **{"topic.record": topic_reference["record"]}),
+                ],
+            ),
+            **kwargs,
+        )
+
+        if results.total == 0:
+            return None
+
+        return next(results.hits)
+
+    def accepted_record(self, identity, record):
+        """Check if current version of record has been accepted."""
+
+        results = self.requests_service.search(
+            identity,
+            extra_filter=dsl.query.Bool(
+                "must",
+                must=[
+                    dsl.Q("term", **{"type": self.request_type_cls.type_id}),
+                    # TODO: make type independent search
+                    dsl.Q("term", **{"topic.record": record["id"]}),
+                    dsl.Q("term", **{"is_open": False}),
+                    dsl.Q("term", **{"status": "accepted"}),
+                    # dsl.Q("range", **{"updated": {"gte": record.updated}}),
+                ],
+            ),
+        )
+        return next(results.hits) if results.total > 0 else None
+
+    @unit_of_work()
+    def create(self, identity, data=None, uow=None, **kwargs):
+        """Creates a RDMCuration request and submits it."""
+
+        role = self.role()
+        assert role, _(
```

Comment:
question: could an `assert` not be removed with an optimization flag? would it not be better to use
```
if not role: 
    raise RuntimeError()
```

or something like that?

Reply:
Good catch. Will update